server/custom_vms.go

```diff
@@ -487,45 +515,116 @@ func (lc *localNetwork) restartNodesWithWhitelistedSubnets(
+		cancel()
+		if err != nil {
+			return err
+		}
+		primaryValidatorsEndtime := make(map[ids.NodeID]time.Time)
+		for _, v := range vs {
+			primaryValidatorsEndtime[v.NodeID] = time.Unix(int64(v.EndTime), 0)
+		}
+		cctx, cancel = createDefaultCtx(ctx)
+		vs, err = platformCli.GetCurrentValidators(cctx, subnetID, nil)
+		cancel()
+		if err != nil {
+			return err
+		}
+		subnetValidators := make(map[ids.NodeID]struct{})
+		for _, v := range vs {
+			subnetValidators[v.NodeID] = struct{}{}
+		}
+		for nodeName, nodeInfo := range nodeInfos {
+			nodeID, err := ids.NodeIDFromString(nodeInfo.Id)
+			if err != nil {
+				return err
+			}
+			_, isValidator := subnetValidators[nodeID]
+			if !isValidator {
+				cctx, cancel := createDefaultCtx(ctx)
+				txID, err := baseWallet.P().IssueAddSubnetValidatorTx(
+					&validator.SubnetValidator{
+						Validator: validator.Validator{
+							NodeID: nodeID,
+							// reasonable delay in most/slow test environments
+							Start: uint64(time.Now().Add(20 * time.Second).Unix()),
+							End:   uint64(primaryValidatorsEndtime[nodeID].Unix()),
+							Wght:  subnetValidatorsWeight,
+						},
+						Subnet: subnetID,
 					},
-					Subnet: subnetID,
-				},
-				common.WithContext(cctx),
-				defaultPoll,
-			)
+					common.WithContext(cctx),
+					defaultPoll,
+				)
+				cancel()
+				if err != nil {
+					return err
+				}
+				zap.L().Info("added the node as a subnet validator",
+					zap.String("subnet-id", subnetID.String()),
+					zap.String("node-name", nodeName),
+					zap.String("node-id", nodeID.String()),
+					zap.String("tx-id", txID.String()),
+				)
+			}
+		}
+	}
+	return nil
+}
+
+func waitSubnetValidators(
+	ctx context.Context,
+	nodeInfos map[string]*rpcpb.NodeInfo,
+	platformCli platformvm.Client,
+	subnetIDs []ids.ID,
+	stopCh chan struct{},
+) error {
+	fmt.Println()
+	color.Outf("{{green}}waiting for the nodes to become subnet validators{{/}}\n")
+	for {
+		notReady := false
+		for _, subnetID := range subnetIDs {
+			cctx, cancel := createDefaultCtx(ctx)
+			vs, err := platformCli.GetCurrentValidators(cctx, subnetID, nil)
 			cancel()
 			if err != nil {
 				return err
 			}
-			zap.L().Info("added the node as a subnet validator",
-				zap.String("subnet-id", subnetID.String()),
-				zap.String("node-id", validatorID.String()),
-				zap.String("tx-id", txID.String()),
-			)
+			subnetValidators := make(map[ids.NodeID]struct{})
+			for _, v := range vs {
+				subnetValidators[v.NodeID] = struct{}{}
+			}
+			for _, nodeInfo := range nodeInfos {
+				nodeID, err := ids.NodeIDFromString(nodeInfo.Id)
+				if err != nil {
+					return err
+				}
+				_, isValidator := subnetValidators[nodeID]
+				if !isValidator {
+					notReady = true
+				}
+			}
+		}
+		if !notReady {
+			return nil
+		}
+		select {
+		case <-stopCh:
+			return errAborted
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-time.After(1 * time.Second):
```

Comment:
Same as above: what does this `1 * time.Second` mean / do? Can we make it a constant?